### PR TITLE
Add --notify flag

### PR DESCRIPTION
**Pull request by bobismijnnaam**
For serious case studies verification can take upwards of 10 minutes. This PR adds the --notify flag that makes waiting a bit more bearable by sending a notification when vercors finishes. Currently supported OSs are mac, ubuntu/debian, and windows 10. No new dependencies are introduced so the added functionality remains lightweight (and should remain so, IMHO - this functionality is not worth including heavyweight OS libraries).

### CODE REVIEW

**Comment by niomaster** [comment]
Are you still planning on changing the mac support?

**Comment by bobismijnnaam** [comment]
I want to discuss tomorrow during the meeting if the clicky behavior of mac notifications is acceptable or not, and also if the swing notifications are too ugly or not. But, this code is not that complex, so if you want to leave out mac support for now so you can merge this today that's fine with me. (The mac section can just be commented and then this PR should be good to go IMO)

**Comment by bobismijnnaam** [comment]
image

Screenshot for how the TrayIcon notification from Swing looks.

Code:

```
import java.awt.*;
import java.awt.event.*;
import java.net.URL;
import javax.swing.*;

public class Main {


  public static void main(String[] a) throws Exception {
    if (SystemTray.isSupported()) {
      Image image = Toolkit.getDefaultToolkit().getImage("images/tray.gif");
      TrayIcon trayIcon = new TrayIcon(image, "Tester2");

      SystemTray tray = SystemTray.getSystemTray();

      try {
        tray.add(trayIcon);
      } catch (AWTException e) {
        System.err.println("TrayIcon could not be added.");
      }

      trayIcon.displayMessage("Tester!", "Some action performed", TrayIcon.MessageType.INFO);

      Thread.sleep(5000);

      System.out.println("bye");
      System.exit(0);
    }
  }
}
```

**Comment by niomaster** [comment]
Can you quickly try with `UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName());`?

**Comment by niomaster** [comment]
I thought I'd just try `SystemTray` myself, but it seems to be unsupported in Ubuntu 20.04 (I guess gnome 3)

**Comment by niomaster** [comment]
Yes, SystemTray support was dropped entirely from gnome 3, but may be enabled again with an addon: https://extensions.gnome.org/extension/1031/topicons/

**Comment by niomaster** [comment]
I recommend against trying this because it adds weird 1 pixel windows and still doesn't look normal with systemlookandfeel

**Comment by bobismijnnaam** [comment]
Too bad, if it is both impractical and ugly we should avoid it. Thanks for looking into it.

**Comment by bobismijnnaam** [comment]
Including the mac look and feel of the above java snippet for completeness:

image

**Comment by bobismijnnaam** [comment]
We discussed this in the meeting and concluded that the current implementation is probably fine. If there is negative feedback about the mac popup's clickability, we'll change the implementation.

File: hre/src/main/java/hre/util/Notifier.scala [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    val os = System.getProperty("os.name").toLowerCase()
+    if (os.contains("win")) {
+      notifyWindows10(title, message)
+    } else if (os.contains("mac")) {
+      notifyMacOS(title, message)
+    } else if (os.contains("nix") || os.contains("linux")) {
+      notifyLibnotify(title, message)
+    } else {
+      false
+    }
```
Comment (niomaster) [inline hre/src/main/java/hre/util/Notifier.scala:22]: this logic deviates a bit from `Configuration` (but seems slightly better), we should standardize it somewhere.

File: hre/src/main/java/hre/util/Notifier.scala [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+      |$ToastXml.LoadXml($ToastTemplate.OuterXml)
+      |
+      |$notify = [Windows.UI.Notifications.ToastNotificationManager]::CreateToastNotifier($app)
+      |
+      |$notify.Show($ToastXml)
+      |""".stripMargin
+
+  def notifyWindows10(title: String, message: String): Boolean = {
+    // Script comes from: https://gist.github.com/Windos/9aa6a684ac583e0d38a8fa68196bc2dc
+    val script = powershellNotificationScript.format(title, message)
```
Comment (niomaster) [inline hre/src/main/java/hre/util/Notifier.scala:80]: this seems suspect, surely you need some escape here?

**Review by niomaster — CHANGES_REQUESTED** [review]

File: hre/src/main/java/hre/util/Notifier.scala (hunk 1 above)
Comment (bobismijnnaam) [reply hre/src/main/java/hre/util/Notifier.scala:22]: I can merge them in a helper method in `Configuration`

File: hre/src/main/java/hre/util/Notifier.scala (hunk 2 above)
Comment (bobismijnnaam) [reply hre/src/main/java/hre/util/Notifier.scala:80]: You mean, escaping "title" and "message" for use in powershell, or escaping the powershellNotificationScript variable? Assuming you mean the second case: VerCors doesn't really have a thread model/security concerns afaik, but if we choose to be defensive I'll add an escaping step.

Comment (niomaster) [reply hre/src/main/java/hre/util/Notifier.scala:80]: I mean the first case, surely if your message is ">" or something it would break.

Comment (bobismijnnaam) [reply hre/src/main/java/hre/util/Notifier.scala:80]: I guess I'll just whitelist "a-zA-Z!? ", that's easy enough and very effective.

File: hre/src/main/java/hre/util/Notifier.scala (hunk 1 above)
Comment (niomaster) [reply hre/src/main/java/hre/util/Notifier.scala:22]: I think I duplicated the logic a few times there, sorry :D

File: hre/src/main/java/hre/util/Notifier.scala (hunk 2 above)
Comment (niomaster) [reply hre/src/main/java/hre/util/Notifier.scala:80]: Oh, that is also fine (did not see your reply in time)

**Review by niomaster — APPROVED** [review]

**Pull request merged**